packages/engine/lib/orchestrator/src/manifest.rs

```diff
@@ -0,0 +1,579 @@
+                            .wrap_err_lazy(|| {
+                                format!("Could not get behavior from dependency: {dependency_name}")
+                            })?
+                    };
+
+                    self.behaviors.push(behavior);
+                }
+                DependencyType::Dataset(_extension) => {
+                    let dataset = get_dataset_from_dependency_projects(
+                        dependency_name,
+                        &dependency_projects,
+                    )?;
+                    self.datasets.push(dataset)
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn add_behavior(&mut self, behavior: SharedBehavior) {
+        self.behaviors.push(behavior);
+    }
+
+    pub fn add_behavior_from_file<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
+        let path = path.as_ref();
+        ensure!(path.is_file(), "Not a behavior file: {path:?}");
+
+        let file_extension = file_extension(&path)?;
+        ensure!(
+            BEHAVIOR_FILE_EXTENSIONS.contains(&file_extension.as_str()),
+            "Not a valid behavior extension: {path:?}"
+        );
+        ensure!(
+            file_extension != "rs",
+            "Custom Rust behaviors are currently unsupported"
+        );
+
+        let file_name = path
+            .file_name()
+            .ok_or_else(|| report!("behavior file expected to have proper file name"))?
+            .to_string_lossy()
+            .to_string();
+        let folder_path = path
+            .parent()
+            .ok_or_else(|| report!("Could not find parent folder for behavior file: {path:?}"))?;
+        let key_path = folder_path.join(&format!("{file_name}.json"));
+
+        self.add_behavior(SharedBehavior {
+            // `id`, `name` and `shortnames` may be updated later if this behavior is a dependency
+            id: file_name.clone(),
+            name: file_name,
+            shortnames: vec![], // if this is a dependency, then these will be updated later
+            behavior_src: file_contents_opt(&path).wrap_err("Could not read behavior")?,
+            // this may not return anything if file doesn't exist
+            behavior_keys_src: file_contents_opt(&key_path)
+                .wrap_err("Could not read behavior keys")?,
+        });
+        Ok(())
+    }
+
+    pub fn add_behaviors_from_directory<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
+        let path = path.as_ref();
+        ensure!(path.is_dir(), "Not a directory: {path:?}");
+
+        debug!("Reading behaviors in {:?}", &path);
+        for entry in path
+            .read_dir()
+            .wrap_err_lazy(|| format!("Could not read behavior directory: {path:?}"))?
+        {
+            match entry {
+                Ok(entry) => {
+                    let path = entry.path();
+                    if BEHAVIOR_FILE_EXTENSIONS.contains(&file_extension(&path)?.as_str()) {
```

Comment:
It seems a bit strange to me that we do this twice:
```rust
ensure!(
    BEHAVIOR_FILE_EXTENSIONS.contains(&file_extension.as_str()),
    "Not a valid behavior extension: {path:?}"
);
```

Reply:
I changed the function, which scans directories for files, to ignore unsuited files.

Reply:
But now we don't get a warning, correct? 
```warn!("Could not add behavior {path:?}: {err:?}");```

Reply:
I just noticed, that the function can also fail, if the file is not readable, but the path is correct. Do you have a suggestion to remove the duplication but keep the warning?

Reply:
I think we should just do `self.add_behavior_from_file` here without the additional check on the extensions, and fail on error.

And then within `add_behavior_from_file` lets
* Check the extension is within `BEHAVIOR_FILE_EXTENSIONS` (and perhaps only warn if it's not)
* **warn** if it's ".rs"
* Fail if we can't read the file, etc.

Reply:
I had that before, but then it will always warn on `"test.js.json"` for behavior keys.

Reply:
I now check for `.json` file extension and otherwise it will simply call `add_behavior_from_file`.